### DISCUSSION TITLE
net.Conn reading race

### DISCUSSION THREAD

**Discussion by maxsupermanhd** [Tnze/go-mc#181] [Q&A] [answered]
So I implemented my proxy and I am trying to parse copies of packets that are passed through (to avoid modifying them by chance) in different gorutine. For some reason sometimes I get straight incorrect data from packets even though id is correct and packet seems to be correct too (not getting connection lost on connected client).

Here is example of how it is breaking itself: (parsing chunk and light packet)
![image](https://user-images.githubusercontent.com/20772987/166747927-6b855759-3f29-422a-87ee-2a8037fdea12.png)
Sources: https://github.com/maxsupermanhd/WebChunk/blob/master/proxy/proxy.go#L245

**Comment by maxsupermanhd** [comment]
I checked with -race flag and it seems like net.Conn.ReadPacket inside bot's HandleGame is still writing to the packet while it is being read by my gorutine that received it through channel in generic event handler, makes no sense for me, how ReadPacket is not blocking until writing operation finished?

**Comment by maxsupermanhd** [comment, marked as answer]
Instead of sending to channel packet structure I decided to reassemble it and make a new slice of bytes, looks like it is working now.

**Comment by Tnze** [comment]
All data is only available in the context which give the data to you. And the buffer will be reused in the next receiving.

You are exporting the data without copying the data, which is not allowed. 

When introducing multithreading and asynchrony, it's your own responsibility to ensure that there is no data race.

**Reply by maxsupermanhd** [reply]
> You are exporting the data without copying the data, which is not allowed.

I thought I am making full object copy, turns out actual slice buffer gets reused, only header is copied. I made sure that I specifically create new buffer with copied data in same thread and reassembled packet: https://github.com/maxsupermanhd/WebChunk/commit/94ac5dfeb7a7e1a718a37116bd8894ae36b063a1

**Comment by Tnze** [comment]
Reusing buffers is a very important optimization method, and allocating memory is a huge overhead, especially for infrastructure such as network library, which needs to pay great attention to performance.

**Reply by maxsupermanhd** [reply]
Since I am sending packet through channel it is difficult to control, if I would've been in C environment maybe I would assemble some sort of a rolling buffer with multithread access that allocated new buffer only if there is a "pile up" but in golang I might just do it this way. Plus there is not a huge amounts of data that I copy, only newly loaded chunks...

**Reply by Tnze** [reply]
I suggest you reconsider whether to introduce goroutine and channel. I don't think it's necessary for you.

**Reply by Tnze** [reply]
is it really necessary to copy the packet?

**Reply by maxsupermanhd** [reply]
The point is that I don't want to block all packet forwarding, in the proxy I am making I also cache packets and do stuff with them. I feel like it should be done in separate thread.

**Reply by Tnze** [reply]
You should always keep the sequence of packet

**Reply by maxsupermanhd** [reply]
They are not processed in parallel, still one processing thread but it must not block packet forwarding at any time. If forwarding will be delayed with processing then it will affect gameplay lag significantly.
